RecoVertex/BeamSpotProducer/plugins/BeamSpotToCUDA.cc

```diff
@@ -0,0 +1,100 @@
+#include <cuda_runtime.h>
+
+#include "CUDADataFormats/BeamSpot/interface/BeamSpotCUDA.h"
+#include "CUDADataFormats/Common/interface/Product.h"
+#include "DataFormats/BeamSpot/interface/BeamSpot.h"
+#include "FWCore/Framework/interface/Event.h"
+#include "FWCore/Framework/interface/MakerMacros.h"
+#include "FWCore/Framework/interface/global/EDProducer.h"
+#include "FWCore/ParameterSet/interface/ConfigurationDescriptions.h"
+#include "FWCore/ParameterSet/interface/ParameterSet.h"
+#include "FWCore/ParameterSet/interface/ParameterSetDescription.h"
+#include "FWCore/ServiceRegistry/interface/Service.h"
+#include "HeterogeneousCore/CUDACore/interface/ScopedContext.h"
+#include "HeterogeneousCore/CUDAServices/interface/CUDAService.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/copyAsync.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/host_noncached_unique_ptr.h"
+
+namespace {
+
+  class BeamSpotHost {
+  public:
+    BeamSpotHost() : data_h_{cms::cuda::make_host_noncached_unique<BeamSpotCUDA::Data>(cudaHostAllocWriteCombined)} {}
+
+    BeamSpotHost(BeamSpotHost const&) = delete;
+    BeamSpotHost(BeamSpotHost&&) = default;
+
+    BeamSpotHost& operator=(BeamSpotHost const&) = delete;
+    BeamSpotHost& operator=(BeamSpotHost&&) = default;
+
+    BeamSpotCUDA::Data* data() { return data_h_.get(); }
+    BeamSpotCUDA::Data const* data() const { return data_h_.get(); }
+
+    cms::cuda::host::noncached::unique_ptr<BeamSpotCUDA::Data>& ptr() { return data_h_; }
+    cms::cuda::host::noncached::unique_ptr<BeamSpotCUDA::Data> const& ptr() const { return data_h_; }
+
+  private:
+    cms::cuda::host::noncached::unique_ptr<BeamSpotCUDA::BeamSpotCUDA::Data> data_h_;
+  };
+
+}  // namespace
+
+class BeamSpotToCUDA : public edm::global::EDProducer<edm::StreamCache<BeamSpotHost>> {
```

Comment:
(Just to better understand the logic) wouldn't a `BemSpotFromCUDA` be eventually needed as well? Or the plan is only to first calculate the beam spot with the cpu, and then move it to CUDA?

Reply:
The beam spot is never reconstructed by the HLT or RECO job, it's simply read from the data stream or from the conditions.
In these cases a `SomethingFromCUDA` module does not make sense.